modules/gitops-argo-cd-round-robin-sharding-algorithm.adoc

```diff
@@ -0,0 +1,138 @@
+
+Using the `round-robin` sharding algorithm in the {gitops-title} provides the following benefits:
+
+* Ensure more balanced workload distribution
+* Prevent shards from being overloaded or underutilized
+* Optimize the efficiency of computing resources
+* Reduce the risk of bottlenecks
+* Improve overall performance and reliability of the Argo CD system
+
+The introduction of alternative sharding algorithms allows for further customization based on specific use cases. You can select the algorithm that best aligns with your deployment needs, which results in greater flexibility and adaptability in diverse operational scenarios.
+
+.Prerequisites
+* You have installed the {gitops-title} Operator in your cluster.
+* You have access to the {OCP} web console.
+* You have access to the cluster with `cluster-admin` privileges.
+
+.Procedure
+
+You can enable the `round-robin` sharding algorithm by either using the {OCP} web console or the command-line interface.
+
+* To enable the `round-robin` algorithm using the {OCP} web console, complete the following steps:
+
+. In the *Administrator* perspective of the web console, go to *Operators* → *Installed Operators*.
+
+. Click *{gitops-title}* from the installed operators and go to the *Argo CD* tab.
+
+. Click the Argo CD instance where you want to enable the `round-robin` sharding algorithm, for example `openshift-gitops`.
+
+. Click the *YAML* tab and edit the yaml file:
++
+.Example Argo CD instance with round-robin sharding algorithm enabled
+[source,yaml]
+----
+apiVersion: argoproj.io/v1beta1
+kind: ArgoCD
+metadata:
+  name: openshift-gitops
+  namespace: openshift-gitops
+spec:
+  controller:
+    processors: {}
+    sharding:
+      enabled: true <1>
+      replicas: 3 <2>
+    env: <3>
+      - name: ARGOCD_CONTROLLER_SHARDING_ALGORITHM 
+        value: round-robin
+----
+<1> Set the `sharding.enabled` parameter to `true` to enable sharding.
+<2> Set the number of replicas to the desired value, for example `3`.
+<3> Set the sharding algorithm to `round-robin`.
+
+. Click *Save* to save the changes.
+
+* To enable the `round-robin` algorithm using the command-line interface, complete the following steps:
+
+. Edit the Argo CD instance YAML file to enable sharding:
++
+[source,terminal]
+----
+$ oc patch argocd <argocd_instance> -n <namespace> --patch='{"spec":{"controller":{"sharding":{"enabled":true}}}}' --type=merge
+----
++
+.Example output
+[source,terminal]
+----
+argocd.argoproj.io/<argocd_instance> patched
+----
+
+. Configure the sharding algorithm to `round-robin`:
++
+[source,terminal]
+----
+$ oc patch argocd <argocd_instance> -n <namespace> --patch='{"spec":{"controller":{"env":[{"name":"ARGOCD_CONTROLLER_SHARDING_ALGORITHM","value":"round-robin"}]}}}' --type=merge
```

Comment:
Hello @Srivaralakshmi I considered this when I was creating this section. My reasoning here is that the three steps do three different things: first enables sharding, second configures round-robin algorithm, and third sets the wanted number of replicas.

What I am thinking here is:
* For step 2, that there is a possibility that there will be more algorithms in the future. Therefore, if I include it separately, it is clear which step the user needs to do to switch between the algorithms.

* For step 3, the user can change the number of replicas to the wanted value. By including the step separately, the user can run the command repeatedly after the whole procedure is done.

This is my reasoning as to why I included it separately, but I am, of course, open to more discussion here. WDYT?

Reply:
@ekristova Well done! Truly appreciate how you have considered the future possibilities while authoring. Excellent trait a writer must have! Thank you for the detailed explanation.

However, I still have this question and possibility in mind:

- Given that the user is expert enough to separate the command and configure the replicas required, can we include future algorithms and replicas in the same step? Will that be beneficial to the user in terms of time and convenience?

Because the output is the same for all 3 steps. We want the writing to be minimalistic and still be beneficial to the user and RH in terms of time, cost, localization efforts, and convenience.

@akram and @varshab1210: We request your review feedback and expert opinion in this situation. Please share your thoughts. Thank you!

Reply:
Hello @ekristova, to add on to what @Srivaralakshmi< I think using one patch command to set all those parameters would be great (See enabling dynamic scaling using CLI)...... Additionally, is see in step 4 you have provided a general command (syntax) for verification, however, you did not add the example command to support the given example output.....

Reply:
@masauso-lungu yes, that is because if I added it there, I would have to add it to all the steps and then all the steps would have general command, example command, and example output, which would be, in my opinion, too crowded. So I rather opted it for the sake of making it more minimalist.

Reply:
But I see, for the sake of consistency, I will change the example output to `<instance_name>-application-controller` examples.

Reply:
> @akram and @varshab1210: We request your review feedback and expert opinion in this situation. Please share your thoughts. Thank you!

Hi all,

I haven't run patched the `argocd` object in one shot to see how the operator's reconciliation loop could handle it. There shouldn't be any issue, the desired stated would be reached, and at worst maybe the controller will be restarted quickly 2 times.

However, as you mentionned for readabilty and reusability of sections in the documentation, maybe having the 3 steps can be helpful for users. There is indeed other sharding algorithm comming, and steps 1 and 3 are common to them. Only step 2 changes.

Reply:
Hello @akram thank you for your input! I was thinking that, to compromise, I could combine steps 1 and 3 into one if they will be the same for other upcoming procedures, and leave the step 2 to change the algorithm separately. I will test if everything will work well and make those changes if you all agree, thank you!